core/trino-main/src/main/java/io/trino/operator/aggregation/listagg/ListaggAggregationFunction.java

```diff
@@ -0,0 +1,428 @@
+                inputParameterMetadata,
+                inputFunction,
+                Optional.empty(),
+                combineFunction,
+                outputFunction,
+                ImmutableList.of(new AccumulatorStateDescriptor(
+                        stateInterface,
+                        stateSerializer,
+                        stateFactory)),
+                outputType);
+
+        GenericAccumulatorFactoryBinder factory = AccumulatorCompiler.generateAccumulatorFactoryBinder(metadata, classLoader);
+        return new InternalAggregationFunction(NAME, inputTypes, ImmutableList.of(intermediateType), outputType, factory);
+    }
+
+    public static void input(Type type, ListaggAggregationState state, Block value, Slice separator, boolean overflowError, Slice overflowTruncationFiller, boolean overflowTruncationCountIndication, int position)
+    {
+        if (state.isEmpty()) {
+            if (overflowTruncationFiller.length() > MAX_TRUNCATION_FILLER_LENGTH) {
+                throw new TrinoException(INVALID_FUNCTION_ARGUMENT, format("truncation filler length %d exceeds maximum length %d", overflowTruncationFiller.length(), MAX_TRUNCATION_FILLER_LENGTH));
+            }
+            // Set the parameters of the LISTAGG command within the state so that
+            // they can be used within the `output` function
+            state.setSeparator(separator);
+            state.setOverflowError(overflowError);
+            state.setOverflowTruncationFiller(overflowTruncationFiller);
+            state.setOverflowTruncationCountIndication(overflowTruncationCountIndication);
+        }
+        state.add(value, position);
+    }
+
+    public static void combine(Type type, ListaggAggregationState state, ListaggAggregationState otherState)
+    {
+        Slice previousSeparator = state.getSeparator();
+        if (previousSeparator == null) {
+            state.setSeparator(otherState.getSeparator());
+            state.setOverflowError(otherState.isOverflowError());
+            state.setOverflowTruncationFiller(otherState.getOverflowTruncationFiller());
+            state.setOverflowTruncationCountIndication(otherState.isOverflowTruncationCountIndication());
+        }
+
+        state.merge(otherState);
+    }
+
+    public static void output(Type type, ListaggAggregationState state, BlockBuilder out)
+    {
+        if (state.isEmpty()) {
+            out.appendNull();
+        }
+        else {
+            outputState(state, out, MAX_OUTPUT_LENGTH);
+        }
+    }
+
+    @VisibleForTesting
+    protected static void outputState(ListaggAggregationState state, BlockBuilder out, int maxOutputLength)
+    {
+        ListaggAggregationOutputContext outputContext = new ListaggAggregationOutputContext(state, maxOutputLength);
+        BlockBuilder aggregationSuffix = VARCHAR.createBlockBuilder(null, 16);
+        state.forEach((block, position) -> appendBlock(block, position, outputContext, out, aggregationSuffix));
```

Comment:
I think this can be dramatically simplified by doing something like:

```java
if (state.isEmpty()) {
    out.appendNull();
}
else {
    long[] length = new long[1];
    boolean[] first = new boolean[] {true};
    state.forEach((block, position) -> {
        if (!first[0]) {
            out.writeBytes(state.getSeparator(), 0, state.getSeparator().length());
            length[0] += state.getSeparator().length();
        }

        first[0] = false;

        int entryLength = block.getSliceLength(position);
        block.writeBytesTo(position, 0, entryLength, out);
        length[0] += entryLength;

        if (length[0] > MAX_OUTPUT_LENGTH) {
            if (state.isOverflowError()) {
                throw new TrinoException(EXCEEDED_FUNCTION_MEMORY_LIMIT, "...");
            }

            out.writeBytes(state.getOverflowTruncationFiller(), 0, state.getOverflowTruncationFiller().length());
            if (state.isOverflowTruncationCountIndication()) {
                out.writeBytes(Slices.utf8Slice("("), 0, 1);
                Slice count = Slices.utf8Slice(Integer.toString(state.getEntryCount()));
                out.writeBytes(count, 0, count.length());
                out.writeBytes(Slices.utf8Slice(")"), 0, 1);
            }

            return false;
        }
    });
    out.closeEntry();
}
```

A few observations:
* There's no actual hard limit for the size of a VARCHAR, so we can take advantage of that and consider MAX_OUTPUT_LENGTH a soft limit. If the limit is exceeded, then we truncate the remainder.
* We should skip adding null entries into the aggregation state. This simplifies the logic for handling the case where all entries are null
* We need to extend `forEach` to allow the consumer to return `false` to short-circuit the iteration.

Reply:
Regarding the simplified code suggestion:
I interpret that the functionality for overflow truncation information in `LISTAGG` should fit within the memory limit.
I did this rather cryptic algorithm in order to make sure that the aggregation with eventual truncation filler and overflow count indication fits within `MAX_OUTPUT_LENGTH`.

In the solution that you are suggesting when dealing with overflows it is likely that the output of the function will be longer than `MAX_OUTPUT_LENGTH`.
Please do advise whether you consider that the function should expose such functionality.


> There's no actual hard limit for the size of a VARCHAR, so we can take advantage of that and consider MAX_OUTPUT_LENGTH a soft limit. If the limit is exceeded, then we truncate the remainder.

If I am understanding correctly, when the remainder is truncated, then the overflow truncation information(positioned at the end of the string) will be truncated.

> We should skip adding null entries into the aggregation state. This simplifies the logic for handling the case where all entries are null
Good point. I've changed the code accordingly.

> We need to extend forEach to allow the consumer to return false to short-circuit the iteration.

`io.trino.operator.aggregation.listagg.SingleListaggAggregationState#forEach` - I don't quite understand what you mean here. In which situation would this be beneficial?

Reply:
> I interpret that the functionality for overflow truncation information in LISTAGG should fit within the memory limit.

That's the intent of what the SQL spec describes, but in Trino there is no hard limit for a single value of type VARCHAR. The limit is driven by other factors such as page size, total size of accumulated values in memory, etc.

> In the solution that you are suggesting when dealing with overflows it is likely that the output of the function will be longer than MAX_OUTPUT_LENGTH.

I think that's ok for now. We can always make it stricter later. The conditions under which it would truncate values are going to be very rare, too (you need to produce a value that's ~1 MB large)

> If I am understanding correctly, when the remainder is truncated, then the overflow truncation information(positioned at the end of the string) will be truncated.

Are you referring to the count being just the count of remaining items? I missed that in my code above. It should subtract how many items it already produced.

> io.trino.operator.aggregation.listagg.SingleListaggAggregationState#forEach - I don't quite understand what you mean here. In which situation would this be beneficial?

We don't need to keep iterating over the remaining items once we decide to truncate. Having forEach short-circuit the loop as soon as the consumer returns `false` would be a way to do it.

Reply:
>Are you referring to the count being just the count of remaining items? I missed that in my code above. It should subtract how many items it already produced.

I'm a bit confused now.
Is the count indication supposed to hold the number of values skipped in the aggregation or the number of bytes skipped in the aggregation. I've first implemented LISTAGG to hold the number of values skipped in the overflow count indication, now it holds the number of bytes skipped. As I understand now, you are suggesting now to hold the number of values.

Reply:
It's supposed to show the number of values skipped. So, for instance, if you have values `a`, `b`, `c`, `d` and `e` and only the first three fit in the output, you should get:

```
a,b,c,...(2)
```

That's what the following line in my suggested code does, except for missing the subtraction of the number of values that were emitted:

```java
Slice count = Slices.utf8Slice(Integer.toString(state.getEntryCount()));
```

It needs to be adjusted to this, and track the number of values emitted so far in the `emitted` variable:

```java
Slice count = Slices.utf8Slice(Integer.toString(state.getEntryCount() - emitted));
```

Reply:
> We need to extend forEach to allow the consumer to return false to short-circuit the iteration.

I decided not to shortcircuit `forEach` method in order to give the possibility of providing a meaningful exception to the user: 

```
throw new TrinoException(EXCEEDED_FUNCTION_MEMORY_LIMIT, format("Concatenated string has the length %d which is larger than the maximum output length %d", outputContext.aggregationBytesLength, maxOutputLength));
```

Reply:
Shortcircuiting would imply also the changing of the logic of the methods: 
- `io.trino.operator.aggregation.AbstractGroupCollectionAggregationState#forEach` 
- `io.trino.operator.aggregation.listagg.SingleListaggAggregationState#forEach`

Overriding the `AbstractGroupCollectionAggregationState#forEach` method for the purpose of `GroupListaggAggregationState` is definitely not straightforward.

Reply:
Indeed. I wouldn't override it -- I would just change the implementation in `AbstractGroupCollectionAggregationState` to allow for that.

You just need to changed it to do:

```java
if (!accept(consumer, values.get(currentBlockId), currentPosition)) {
    return;
}
```